src/passes/OptimizeCasts.cpp

```diff
@@ -215,22 +463,35 @@ struct OptimizeCasts : public WalkerPass<PostWalker<OptimizeCasts>> {
       return;
     }
 
-    // First, find the best casts that we want to use.
+    // Look for casts which can be moved earlier
+    EarlyCastFinder earlyCastFinder(getPassOptions(), getModule(), func);
+    earlyCastFinder.walkFunctionInModule(func, getModule());
+
+    bool castsToMove = earlyCastFinder.hasCastsToMove();
+    if (castsToMove) {
+      // Duplicate casts to earlier locations if possible
+      EarlyCastApplier earlyCastApplier(earlyCastFinder);
+      earlyCastApplier.walkFunctionInModule(func, getModule());
+
+      // Adding more casts causes types to be refined, that should be
+      // propagated. Especially those of nested casts.
```

Comment:
I meant that nested casts would cause at least one of the casts types to become potentially inaccurate. But since now that I think of it, it happens to any expression, I should just remove this as you suggest.